### ISSUE TITLE
Image based streamlines_registration: unable to warp streamlines into template space

### ISSUE THREAD

**Issue by WilliamFCB** [dipy/dipy#2703]
Hi All,

I tried to warp streamlines into 1mm FA template space, by adapting the code in the streamlines_registration.py
I seem to have successfully calculated the necessary deformation field (see page 1 of attached streamline_registration.pdf), but applying this to the streamlines ( in this case left CST) does not give the expected result: See page 2 of the pdf, which shows the  "warped" streamlines and the underlying FA template) and pages 3 and 4, showing respectively the warped streamlines overlaid  on the FA_template and the warped subject FA image in mrview. As shown, on page 5 of the pdf, the warped streamlines seem to "perfectly" fit the "raw" FA image, indicating something went wrong in applying the deformation field. 
I am not sure if this is related to the issue reported in #2400

I attached my code at the end.
Any suggestions are very much appreciated!

Many thanks in advance
William

Data:
DWI data was preprocessed with qsiprep (e.g., gibs ringing correction, intensity normalisation, denoiseing, topup and eddy correction (including correction for movement by susceptibly  interaction). TractSeg was used to extract fiberbundles (tck, 10000). Before running TractSeg, data was rigidly registered to the FSL_HCP1065_FA_1mm.nii.gz template.

Background:
The reason for wanting to non-linearly warp streamlines, was the observation that when comparing Tractometry tract segment profiles (as implemented in TractSeg, which is based on the Bundle Analytics approach) across hundreds of subjects, suggested that unaccounted for anatomical variation might hamper comparing segment profiles across subjects. My idea is to warp both the streamlines and relevant DWI images and do tractometry on these.

Specifically, to compare profiles across subjects, we were interested to investigate, which segments were most comparable across subjects and ignore segments that display large inter subject variation (e.g. dispersion)

To this end we calculated the tensor for each segment and extracted the Eigenvalues, to see for which segments Eigen Value 1 was close to 1.

To calculate the eigenvalues, I used the following procedure:

- I split generated TOM/tract.nii.gz images in their X,Y,Z components using fslsplit
   TOMs (Tract Orientation Maps contain one peak per voxel representing the main orientation of the respective bundle at this voxel)
- I calculated: XX, XY, XZ, YY, YZ, and ZZ using fslmaths
- Ran Tractometry on these for all tracts.
- Created for each tract a csv file with 9 columns: XX, XY, XZ, XY, YY, YZ, XZ, YZ, ZZ and 98 rows (segments)
- For each segment these were subsequently reshaped to a 3x3 tensor matrix, from which I extracted eigenvalues and eigenvectors

I link one file displaying the EigenValue 1 for al tracts for a cohort of 157 subjects (childeren ~ 11 years of age):
https://www.dropbox.com/s/bti78gzcajktkcs/cfin_tensor_EigVal1_ALL-Tracts_ALL-subjects_tck_10000.pdf?dl=0

Additionally, I attach  results for the left Uncinate Fasciculus, with tractometry profiles for FA, MD, RD, L1, Eigen values 1,2 and 3, and Mode, containing data from two Cohorts: CFIN (N=157) and DRCMR (N = 143), age ~11 years

As you can see, there seems to be quite some variation across subjects. Through the noise, it appears that while Eigen Value 1 is generally of the same shape across subjects, hills and valleys do not overlap. To us this suggests that there is unaccounted anatomical variation, and that segments as such are thus not necessarily comparable across subjects.

Sytem info:
OS: Linux-5.14.0-162.6.1.el9_1.x86_64-x86_64-with-glibc2.10
Python: Python 3.8.13
DIPY: 1.5.0

```python
##########################################
Used Code
##########################################

#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Dec 16 08:50:24 2022
adapted from DIPY streamlines_registration.py
@author: wimb
"""
"""
================================================
Applying image-based deformations to streamlines
================================================

"""

import numpy as np
import nibabel as nib
import os
from dipy.io.image import load_nifti_data, load_nifti, save_nifti


"""
set output path for saving overlayimages
"""

imdir="/mnt/scratch/VIA11/sub-via003ses01prisma/streamline_registration_images"
if os.path.exists(imdir):
    print(imdir, ' : Exits')
else:
    print(imdir, ' : Does not Exists, creating it')
    os.mkdir(imdir)                 # create images sub directory
    os.chmod(imdir, 0o750)          # set permissions to rwxrw....


"""
Here we calculate the non-linear registration by  warping the raw brainmasked FA in MNI orientation
to the FSL_HCP1065_FA_1mm.nii.gz template. The subject FA is rigidly aligened to the FA_template.
Both are in 1x1x1 isotropic voxels.
"""


"""
load the FA_template and subject FA data
"""
fa_temp_name="/mnt/projects/VIA11/MRI/Scripts/wimb/Tract_register/FSL_HCP1065_FA_1mm.nii.gz"
fa_temp_data, fa_temp_affine, fa_temp_img = load_nifti(fa_temp_name, return_img=True)


fa_raw_name="/mnt/scratch/VIA11/sub-via003ses01prisma/sub-via003ses01prisma_dwi_MNI_FA.nii.gz"
fa_raw_data, fa_raw_affine, fa_raw_img = load_nifti(fa_raw_name, return_img=True)
vox_size = fa_raw_img.header.get_zooms()[0]

"""
We read the streamlines from file in voxel space:

"""
from dipy.io.streamline import load_tractogram

tr="/mnt/scratch/VIA11/sub-via003ses01prisma/tractseg_output/TOM_trackings_tck_10000/CST_left.tck"
sft = load_tractogram(tr, fa_raw_img)

"""
We will first perform an affine registration to roughly align the two volumes:

"""

from dipy.align.imaffine import (MutualInformationMetric, AffineRegistration,
                                 AffineMap, transform_origins)
from dipy.align.transforms import (TranslationTransform3D, RigidTransform3D,
                                   AffineTransform3D)
from dipy.viz import regtools

static = np.asarray(fa_temp_img.dataobj)
static_affine = fa_temp_img.affine
moving = np.asarray(fa_raw_img.dataobj)
moving_affine = fa_raw_img.affine


"""
Create overlay of static and moving to show initial fit
"""
identity = np.eye(4)

#As the images have the same resolution there is no need to resample the moving one
regtools.overlay_slices(static, moving, None, 0,
                        "Static", "Moving", imdir + "/initial_fit_sagittal.png",dpi = 900)
regtools.overlay_slices(static, moving, None, 1,
                        "Static", "Moving", imdir + "/initial_fit_coronal.png",dpi = 900)
regtools.overlay_slices(static, moving, None, 2,
                        "Static", "Moving", imdir + "/initial_fit_axial.png",dpi = 900)

"""
We estimate an affine that maps the origin of the moving image to that of the
static image. We can then use this later to account for the offsets of each
image.

"""

affine_map = transform_origins(static, static_affine, moving, moving_affine)

"""
We specify the mismatch metric:

"""

nbins = 32
sampling_prop = None
metric = MutualInformationMetric(nbins, sampling_prop)

"""
As well as the optimization strategy:

"""

level_iters = [10, 10, 5]
sigmas = [3.0, 1.0, 0.0]
factors = [4, 2, 1]
affine_reg = AffineRegistration(metric=metric, level_iters=level_iters,
                                sigmas=sigmas, factors=factors)
transform = TranslationTransform3D()

params0 = None
translation = affine_reg.optimize(static, moving, transform, params0,
                                  static_affine, moving_affine)
transformed = translation.transform(moving)
transform = RigidTransform3D()

rigid_map = affine_reg.optimize(static, moving, transform, params0,
                                static_affine, moving_affine,
                                starting_affine=translation.affine)
transformed = rigid_map.transform(moving)
transform = AffineTransform3D()

"""
We bump up the iterations to get a more exact fit:

"""

affine_reg.level_iters = [1000, 1000, 100]
highres_map = affine_reg.optimize(static, moving, transform, params0,
                                  static_affine, moving_affine,
                                  starting_affine=rigid_map.affine)
transformed = highres_map.transform(moving)


"""
Create overlay of static and moving to show high res affine fit
"""
identity = np.eye(4)

#As the images have the same resolution there is no need to resample the moving one
regtools.overlay_slices(static, transformed, None, 0,
                        "Static", "Moving", imdir + "/highres_affine_sagittal.png",dpi = 900)
regtools.overlay_slices(static, transformed, None, 1,
                        "Static", "Moving", imdir + "/highres_affine_coronal.png",dpi = 900)
regtools.overlay_slices(static, transformed, None, 2,
                        "Static", "Moving", imdir + "/highres_affine_axial.png",dpi = 900)


"""
We now perform the non-rigid deformation using the Symmetric Diffeomorphic
Registration (SyN) Algorithm proposed by Avants et al. [Avants09]_ (also
implemented in the ANTs software [Avants11]_):

"""

from dipy.align.imwarp import SymmetricDiffeomorphicRegistration
from dipy.align.metrics import CCMetric

metric = CCMetric(3)
level_iters = [10, 10, 5]
sdr = SymmetricDiffeomorphicRegistration(metric, level_iters)

mapping = sdr.optimize(static, moving, static_affine, moving_affine,
                       highres_map.affine)
warped_moving = mapping.transform(moving)


"""
Save a warped version of the warped moving FA. for checking
At this point one can also output other diffusivity images (e.g. MD,L1,RD)
for tractometry latter on

"""
outMNI="/mnt/scratch/VIA11/sub-via003ses01prisma/tractseg_output/MNI_FA.nii"
save_nifti(outMNI, warped_moving, fa_temp_img.affine, hdr=None)


"""
Create overlay of static and moving to show diffeo warped fit:

"""
from dipy.viz import regtools

regtools.overlay_slices(static, warped_moving, None, 0, 'Static', 'Moving',
                        imdir + '/diffeo_warped_sagittal.png',dpi = 900)
regtools.overlay_slices(static, warped_moving, None, 1, 'Static', 'Moving',
                        imdir + '/diffeo_warped_coronal.png', dpi = 900)
regtools.overlay_slices(static, warped_moving, None, 2, 'Static', 'Moving',
                        imdir + '/diffeo_warped_axial.png', dpi = 900)

"""
.. figure:: transformed_sagittal.png
   :align: center
.. figure:: transformed_coronal.png
   :align: center
.. figure:: transformed_axial.png
   :align: center

   Deformable registration result.

"""

"""
We then apply the obtained deformation field to the streamlines. Note that the
process can be sensitive to image orientation and voxel resolution. Thus, we
first apply the non-rigid warping and simultaneously apply a computed rigid
affine transformation whose extents must be corrected to account for the
different voxel grids of the moving and static images.

"""

from dipy.tracking.streamline import deform_streamlines

# Create an isocentered affine
target_isocenter = np.diag(np.array([-vox_size, vox_size, vox_size, 1]))

# Take the off-origin affine capturing the extent contrast between the mean b0
# image and the template
origin_affine = affine_map.affine.copy()

"""
In order to align the FOV of the template and the mirror image of the
streamlines, we first need to flip the sign on the x-offset and y-offset so
that we get the mirror image of the forward deformation field.

We need to use the information about the origin offsets (i.e. between the
static and moving images) that we obtained using :meth:`transform_origins`:

"""

origin_affine[0][3] = -origin_affine[0][3]
origin_affine[1][3] = -origin_affine[1][3]

#"""
#:meth:`transform_origins` returns this affine transformation with (1, 1, 1)
#zooms and not (2, 2, 2), which means that the offsets need to be scaled by 2.
#Thus, we scale z by the voxel size:

#"""

#origin_affine[2][3] = origin_affine[2][3]/vox_size

#"""
#But when scaling the z-offset, we are also implicitly scaling the y-offset as
#well (by 1/2).Thus we need to correct for this by only scaling the y by the
#square of the voxel size (1/4, and not 1/2):
#
#"""

#origin_affine[1][3] = origin_affine[1][3]/vox_size**2

# Apply the deformation and correct for the extents

mni_streamlines = deform_streamlines(
    sft.streamlines, deform_field=mapping.get_forward_field(),
    stream_to_current_grid=target_isocenter,
    current_grid_to_world=origin_affine, stream_to_ref_grid=target_isocenter,
    ref_grid_to_world=np.eye(4))


"""
We display the original streamlines and the registered streamlines:

"""

from dipy.viz import has_fury

outname=imdir + "/MNI_CST_left_tck.png"
def show_template_bundles(bundles, show=False, fname= outname):

    scene = window.Scene()
    template_actor = actor.slicer(static)
    scene.add(template_actor)

    lines_actor = actor.streamtube(bundles, window.colors.orange,
                                   linewidth=0.3)
    scene.add(lines_actor)

    if show:
        window.show(scene)
    if fname is not None:
        window.record(scene, n_frames=1, out_path=fname, size=(900, 900))


if has_fury:

    from fury import actor, window

    show_template_bundles(mni_streamlines, show=False,
                          fname=outname)

"""
Finally, we save the registered streamlines:

"""

from dipy.io.stateful_tractogram import Space, StatefulTractogram
from dipy.io.streamline import save_tractogram

tractmni = StatefulTractogram(mni_streamlines, fa_temp_img, Space.RASMM)

save_tractogram(tractmni, '/mnt/scratch/VIA11/sub-via003ses01prisma/tractseg_output/MNI_CST_left.tck', bbox_valid_check=True)

```

[UF_left_tck_10000_ALL_subjects.pdf](https://github.com/dipy/dipy/files/10278156/UF_left_tck_10000_ALL_subjects.pdf)
[Streamline_registration.pdf](https://github.com/dipy/dipy/files/10278159/Streamline_registration.pdf)

**Comment by skoudoro** [comment]
Hi @WilliamFCB,

Thank you for this complete report.

Yes it is related to #2400 and #2369 should fix it. We need to update this "Applying image-based deformations to streamlines" tutorial which contains mistake. it will be done ASAP.

**Comment by WilliamFCB** [comment]
Hi @skoudoro,
I hope you are doing fine. I was wondering if the issue has been resolved in DIPY 1.6.
Cheers
